### PR TITLE
PNT 22 other polling sources

### PR DESCRIPTION
**Pull request by gunsmoke**
All Sources Call function now returns a Response and have Backoff implemented

~~TODO: implement a simple weight system to choose witch Sources to use~~

### CODE REVIEW

**Comment by PaulBernier** [comment]
OpenExchangeRates gets its data from the European Central Bank, which publishes the rates only once a day. I don't think we should poll that data source given our use case.

**Comment by gunsmoke** [comment]
@PaulBernier Got it, I will remove it then

PS: ExchangeRatesAPI is the same right?

Maybe we can leave `ExchangeRatesAPI` just for Local testing? ( no key required )

**Comment by PaulBernier** [comment]
Oooops. I actually meant `ExchangeRatesAPI` update only once a day and not `OpenExchangeRates`which  can give much more precise data.

We can keep `ExchangeRatesAPI` for testing if you want but we should make sure that in prod it cannot be used. Users won't bother researching the quality of the pre-configured data sources and this could skew the values if they inadvertently pick this one.

**Comment by MitchellBerry** [comment]
At this stage different polling sources seem incredibly useful. Without it the grading mechanism isn't actually being tested, it's all down to hash difficulty. The current testnet exchange rate source requires a paid subscription to contribute to testnet, otherwise the free api limit runs out in a few days, this restricts developer contributions.

As a final outcome you would guess a miner wants complete control over the Oracle values they submit.
For many using general inbuilt sources is likely fine, offering more options doesn't hurt at all.

 In a highly competitive mining environment peg values should be able to be updated at will to anything the miner decides is most likely to win.

File: polling/exchangeratesapi.go [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+	"net/http"
+)
+
+type ExchangeRatesAPIResponse struct {
+	Date    string
+	Base    string
+	Rates   ExchangeRatesAPIRecord
+}
+
+type ExchangeRatesAPIRecord struct {
```
Comment (Emyrk) [inline polling/exchangeratesapi.go]: Can we change this to a map of floats? This just seems klugy.
See what I mean in this playground.
https://play.golang.org/p/Ynmo2ZSC4fj

Then we can write the list of currencies we want, and do it all dynamically.

File: polling/exchangeratesapi.go [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	peg.CHF.Value = Round(response.Rates.CHF)
+	peg.CHF.When = timestamp
+	peg.INR.Value = Round(response.Rates.INR)
+	peg.INR.When = timestamp
+	peg.SGD.Value = Round(response.Rates.SGD)
+	peg.SGD.When = timestamp
+	peg.CNY.Value = Round(response.Rates.CNY)
+	peg.CNY.When = timestamp
+	peg.HKD.Value = Round(response.Rates.HKD)
+	peg.HKD.When = timestamp
```
Comment (Emyrk) [inline polling/exchangeratesapi.go]: If we do a map of strings above, we can make this a loop over some constant list vs all this repeated stuff.

File: polling/openexchangerates.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+	XDR float64
+	XOF float64
+	XPD float64
+	XPF float64
+	XPT float64
+	YER float64
+	ZAR float64
+	ZMW float64
+	ZWL float64
+}
```
Comment (Emyrk) [inline polling/openexchangerates.go]: See my comment above about making these things a map

File: polling/assets.go [hunk 4]
```diff
@@ -109,113 +166,44 @@ func PullPEGAssets(config *config.Config) (pa PegAssets) {
...
-		Peg.INR.Value = Round(APILayerResponse.Quotes.USDINR)
-		Peg.INR.When = string(APILayerResponse.Timestamp)
-		Peg.SGD.Value = Round(APILayerResponse.Quotes.USDSGD)
-		Peg.SGD.When = string(APILayerResponse.Timestamp)
-		Peg.CNY.Value = Round(APILayerResponse.Quotes.USDCNY)
-		Peg.CNY.When = string(APILayerResponse.Timestamp)
-		Peg.HKD.Value = Round(APILayerResponse.Quotes.USDHKD)
-		Peg.HKD.When = string(APILayerResponse.Timestamp)
+	
+	digital_currencies, currency_rates, precious_metals := GetAvailableAssetsByWieght(config)
```
Comment (Emyrk) [inline polling/assets.go]: Typo, `Wieght` -> `Weight`

Comment (gunsmoke) [reply polling/assets.go]: yikes, Nice catch.

**Pull request merged**